src/renku_crac/app.py

```diff
@@ -0,0 +1,369 @@
+"""Compute resource access control (CRAC) app."""
+from dataclasses import dataclass
+from typing import Type, cast
+
+from pydantic import EmailStr
+from sanic import HTTPResponse, Request, Sanic, json
+from sanic_ext import validate
+
+from src import models
+from src.models import errors
+from src.renku_crac.config import Config
+from src.renku_crac.error_handler import CustomErrorHandler
+from src.schemas import apispec
+
+
+@dataclass
+class Server:
+    """Server contains all handlers for CRAC and the configuration."""
+
+    config: Config
+
+    def _get_resource_pool(self):
+        async def handler(request: Request, resource_pool_id: int):
+            res = await self.config.db.get_resource_pools(resource_pool_id, name=request.args.get("name"))
+            if len(res) < 1:
+                raise errors.MissingResourceError(
+                    message=f"The resource pool with id {resource_pool_id} cannot be found."
+                )
+            return json(apispec.ResourcePoolWithId.from_orm(res[0]).dict(exclude_none=True))
+
+        return handler
+
+    def _get_resource_pools(self):
+        async def handler(_: Request):
+            res = await self.config.db.get_resource_pools()
+            return json([apispec.ResourcePoolWithId.from_orm(r).dict(exclude_none=True) for r in res])
+
+        return handler
+
+    def _post_resource_pools(self):
+        @validate(json=apispec.ResourcePool)
+        async def handler(_: Request, body: apispec.ResourcePool):
+            rp = models.ResourcePool.from_dict(body.dict())
+            res = await self.config.db.insert_resource_pool(rp)
+            return json(apispec.ResourcePoolWithId.from_orm(res).dict(exclude_none=True), 201)
+
+        return handler
+
+    def _post_user(self):
+        @validate(json=apispec.UserWithId)
+        async def handler(_: Request, body: apispec.UserWithId):
+            users = await self.config.db.get_users(keycloak_id=body.id)
+            if len(users) == 1:
+                # The user already exists in the crac db
+                user = users[0]
+                if user.username != body.username:
+                    # The username in keycloak is different, so update the user to have the username from keycloak
+                    user = await self.config.db.update_user(keycloak_id=cast(int, user.id), username=body.username)
+                return json(
+                    apispec.UserWithId(id=user.keycloak_id, username=cast(EmailStr, user.username)).dict(
+                        exclude_none=True
+                    ),
+                    200,
+                )
```

Comment:
I think this is overly strict and inconvenient for whoever is using the API. Especially for the users endpoint. Here the POST is idempotent because we cannot have duplicates for the user keycloak id or username. So if someone wants to create the same user that is already in the db I can simply return it. There is no reason to error out when the state that the caller wants is already there.

I just think it is really weird and unfair to send a 409 and force the caller to send another GET request when the state they want the db to be in is already there.

Reply:
fair enough.

Another thing in the code above that confuses me is that if the user doesn't exist and the username doesn't match the keycloak username, we error out. but if the user already exists and the names don't match, we set the name to the supplied name, contradicting the comment "update the user to have the username from keycloak". Right now it's "update the user to have the username from the request".
And if the names need to match, do we even need to pass `username` in the request, if all that matters is the keycloak `id`?

Reply:
Ah yeah that is a very good point. Sorry I think I got lost in my own logic there. Let me fix this.

Reply:
So after looking at this again and thinking in general I think it is better to remove the username. We are just asking for trouble because we can easily get into a situation where the username in our db is different form keycloak for the same user. And then we can have all sorts of unpredictable behaviour. So until we are certain we need to have the usernames from keycloak in our db it is better to simply leave them out.

Reply:
So I removed the username field from our db and also simplified the logic you were talking about. I think it is now clearer. But let me know I can modify or clarify thing further.